rmw_connext_shared_cpp/include/rmw_connext_shared_cpp/wait.hpp

```diff
@@ -15,15 +15,103 @@
 #ifndef RMW_CONNEXT_SHARED_CPP__WAIT_HPP_
 #define RMW_CONNEXT_SHARED_CPP__WAIT_HPP_
 
+#include <unordered_map>
+#include <unordered_set>
+
 #include "ndds_include.hpp"
 
 #include "rmw/error_handling.h"
 #include "rmw/impl/cpp/macros.hpp"
 #include "rmw/types.h"
 
 #include "rmw_connext_shared_cpp/condition_error.hpp"
+#include "rmw_connext_shared_cpp/event_converter.hpp"
 #include "rmw_connext_shared_cpp/types.hpp"
 #include "rmw_connext_shared_cpp/visibility_control.h"
+#include "rmw_connext_shared_cpp/connext_static_event_info.hpp"
+
+
+rmw_ret_t __gather_event_conditions(
+  rmw_events_t * events,
+  std::unordered_set<DDS::StatusCondition *> & status_conditions)
+{
+  RMW_CHECK_ARGUMENT_FOR_NULL(events, RMW_RET_INVALID_ARGUMENT);
+
+  std::unordered_map<DDS::StatusCondition *, DDS::StatusMask> status_mask_map;
+  // gather all status conditions and masks
+  for (size_t i = 0; i < events->event_count; ++i) {
+    rmw_event_t * current_event = static_cast<rmw_event_t *>(events->events[i]);
+    DDS::Entity * dds_entity = static_cast<ConnextCustomEventInfo *>(
+      current_event->data)->get_entity();
```

Comment:
What if `current_event->data` is null?

Reply:
Since this is technically an invalid argument, I'll add an invalid argument check unless we want a better error message

Reply:
That's fine by me, thanks :)